File: music/src/main/java/moe/kyokobot/music/MusicQueue.java
package moe.kyokobot.music;

import com.sedmelluq.discord.lavaplayer.track.AudioTrack;
import lombok.Getter;
import lombok.Setter;
import moe.kyokobot.bot.command.CommandContext;
import moe.kyokobot.bot.util.StringUtil;
import net.dv8tion.jda.core.entities.Guild;
import net.dv8tion.jda.core.entities.TextChannel;

import java.util.Collections;
import java.util.LinkedList;

import static java.lang.String.format;
import static moe.kyokobot.music.MusicIcons.PLAY;

public class MusicQueue {
    @Getter
    private final MusicManager manager;

    @Getter
    private final Guild guild;

    private TextChannel announcingChannel;

    private CommandContext context;

    @Getter
    private LinkedList<AudioTrack> tracks;

    @Getter
    private AudioTrack lastTrack;

    @Getter
    @Setter
    private boolean repeating;

    public MusicQueue(MusicManager manager, Guild guild) {
        this.manager = manager;
        this.guild = guild;
        tracks = new LinkedList<>();
    }

    public void add(AudioTrack track) {
        tracks.add(track);
    }

    public void clear() {
        tracks.clear();
    }

    public AudioTrack poll() {
        return tracks.isEmpty() ? null : (lastTrack = tracks.removeFirst());
    }

    public void announce(AudioTrack track) {
        if (announcingChannel != null) {
            announcingChannel.sendMessage(PLAY + format(context.getTranslated("music.nowplaying"),
                    track.getInfo().title.replace("`", "\\`"),
                    StringUtil.musicPrettyPeriod(track.getDuration()))).queue();
        }
    }

    public boolean isEmpty() {
        return tracks.isEmpty();
    }

    public void shuffle() {
        Collections.shuffle(tracks);
    }

    public void remove(int index) {
        if (tracks.size() >= index) return;
        tracks.remove(index);
    }

    public void setAnnouncing(TextChannel announcingChannel, CommandContext context) {
        this.announcingChannel = announcingChannel;
        this.context = context;
    }
}

File: music/src/main/java/moe/kyokobot/music/MusicQueue.kt
package moe.kyokobot.music

import com.sedmelluq.discord.lavaplayer.track.AudioTrack
import moe.kyokobot.bot.command.CommandContext
import moe.kyokobot.bot.util.StringUtil
import net.dv8tion.jda.core.entities.Guild
import net.dv8tion.jda.core.entities.Member
import net.dv8tion.jda.core.entities.TextChannel
import java.util.*
import java.util.function.Predicate

class MusicQueue(val manager: MusicManager, val guild: Guild) {
    var lastTrack: AudioTrackWrapper? = null
        get() = if (tracks.isEmpty()) null else tracks.removeFirst()
        private set

    var binder: Member? = null
    var boundChannel: TextChannel? = null
    var announcingChannel: TextChannel? = null
        get() = context?.channel
    var context: CommandContext? = null

    var repeating = false
    val tracks = LinkedList<AudioTrackWrapper>()

    fun clear() = tracks.clear()
    fun poll() = lastTrack
    fun shuffle() = tracks.shuffle()
    fun add(track: AudioTrackWrapper) {
        if (tracks.size < 250)
            tracks.add(track)
    }
    fun isEmpty() = tracks.isEmpty()
    fun remove(index: Int) {
        if (index >= tracks.size) return
        tracks.removeAt(index)
    }
    fun removeUser(user: String) {
        tracks.removeIf { track -> track.user == user }
    }

    fun removeDuplicates() {
        for(i in 1 until tracks.size) {
           if(tracks.get(i).audioTrack.equals(tracks.get(i - 1).audioTrack)) {
               tracks.removeAt(i);
           }
        }
    }

    fun announce(track: AudioTrackWrapper) {
        val channel = if (boundChannel != null) boundChannel else announcingChannel
        channel?.sendMessage(MusicIcons.PLAY + context?.getTranslated("music.nowplaying")?.format(
                track.audioTrack.info.title.replace("`", "\\`"),
                StringUtil.musicPrettyPeriod(track.audioTrack.duration)))?.queue()
    }
}